File: strategies/cargo.go
// Copyright 2025 Naren Yellavula
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package strategies

// CargoHelpStrategy handles Cargo commands
type CargoHelpStrategy struct {
	cmdRunner *CommandRunner
}

func NewCargoHelpStrategy(cmdRunner *CommandRunner) *CargoHelpStrategy {
	return &CargoHelpStrategy{cmdRunner: cmdRunner}
}

func (c *CargoHelpStrategy) SupportsCommand(baseCmd string) bool {
	return baseCmd == "cargo"
}

func (c *CargoHelpStrategy) Priority() int {
	return 2
}

func (c *CargoHelpStrategy) GetHelp(cmdParts []string) (string, error) {
	cmd := NewCommand(cmdParts)

	if !cmd.HasSubCommand(1) {
		return c.cmdRunner.Run("cargo", "--help")
	}

	subCmd := cmd.GetSubCommand(0)
	return c.cmdRunner.Run("cargo", subCmd, "--help")
}

File: recaller-rs/src/help_system/strategies/cargo.rs
use std::sync::Arc;

use anyhow::Result;

use crate::help_system::runner::CommandRunner;
use crate::help_system::strategy::{CommandParts, HelpStrategy};

pub struct CargoHelpStrategy {
    runner: Arc<CommandRunner>,
}

impl CargoHelpStrategy {
    pub fn new(runner: Arc<CommandRunner>) -> Self {
        Self { runner }
    }
}

impl HelpStrategy for CargoHelpStrategy {
    fn priority(&self) -> i32 {
        2
    }

    fn supports_command(&self, base_cmd: &str) -> bool {
        base_cmd == "cargo"
    }

    fn get_help(&self, command: &CommandParts) -> Result<String> {
        if !command.has_sub_command(1) {
            return self.runner.run("cargo", &["--help"]);
        }

        let sub_cmd = command.get_sub_command(0).unwrap();
        self.runner.run("cargo", &[sub_cmd, "--help"])
    }
}
